### DISCUSSION TITLE
WinUI 3 Access LocalState Folder

### DISCUSSION THREAD

**Discussion by Bondarenko1990** [microsoft/WindowsAppSDK#2581] [Dev team discussions]
I use in my application Sqlite. 
![image](https://user-images.githubusercontent.com/32612104/172071839-102650c5-3be7-4f14-9c44-ef55361110ca.png)

And I want to add migrations. But....
**The EF migration command uses dotnet build, which is not supported by WinUI project now.
The workaround is defining EF related things in another project and reference it from WinUI project.** 
https://github.com/microsoft/microsoft-ui-xaml/issues/3850#issuecomment-750898842
And how can I get the path to the database in class library project? 
![image](https://user-images.githubusercontent.com/32612104/172071907-a441b8e1-3c1a-4120-b6d1-de254f23667e.png)

**Comment by castorix** [comment]
To get special folders, you can use _Environment.GetFolderPath_, like :

` string sLocalAppdata = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);`

**Reply by Bondarenko1990** [reply]
If I use Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData) I get path (C:\****\bonds\AppData\Local) and the database is created in the folder LocalCache.
I need LocalSatet Folder of my App (C:\Users\****\AppData\Local\Packages\******_c36cftgf3se56\LocalState) not (C:\Users\****\AppData\Local\Packages\******_c36cftgf3se56\LocalCache).

**Comment by huoyaoyuan** [comment]
You can target the library project to `net6.0-windows10.0.19041` (or any other SDK version), without adding WinUI, so that you are supported by `dotnet build` and have access to `ApplicationData` which is a part of WinRT api.

**Reply by Bondarenko1990** [reply]
I also have migration errors
![image](https://user-images.githubusercontent.com/32612104/172218946-6696d713-8203-4364-ae50-c10164ed9d8d.png)

**Reply by Bondarenko1990** [reply]
after adding the CsWinRT library
```
<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
   	 <TargetFramework>net5.0-windows10.0.19041.0</TargetFramework>
	   <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
	   <Platforms>x64;x86;arm64</Platforms>
   </PropertyGroup>
	<!-- CsWinRT properties -->
    <PropertyGroup>
    	<CsWinRTComponent>true</CsWinRTComponent>
    	<CsWinRTWindowsMetadata>10.0.19041.0</CsWinRTWindowsMetadata>
    	<OutputType>Library</OutputType>
    	<!-- Uncomment the line below to generate a log.txt in the obj folder -->
    	<!--<CsWinRTEnableLogging>true</CsWinRTEnableLogging>-->
    </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="5.0.10" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="5.0.10" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="5.0.10">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.Toolkit.Mvvm" Version="7.1.0" />
    <PackageReference Include="Microsoft.Windows.CsWinRT" Version="1.6.4" />
    <PackageReference Include="System.Runtime.InteropServices.WindowsRuntime" Version="4.3.0" />
  </ItemGroup>
</Project>
```

I also have migration errors
![image](https://user-images.githubusercontent.com/32612104/172243391-cb097e50-8684-403b-8b72-deb0e1b2ce63.png)

**Reply by huoyaoyuan** [reply]
Don't use package manager. Try the `dotnet ef` command line tool.
Never add CsWinRT reference directly.

**Reply by Bondarenko1990** [reply]
I tried... no good results
Maybe I'm doing something wrong ... can you show examples?

**Reply by Bondarenko1990** [reply]
@huoyaoyuan Have you tried making a migration in the CsWinRT project?

**Reply by huoyaoyuan** [reply]
No, I never set connection string directly in `OnConfigure`.
I remember in fact you can omit the application data folder. It will be your current working directory in packaged app.

**Reply by Bondarenko1990** [reply]
@huoyaoyuan What does it mean to omit?

**Reply by huoyaoyuan** [reply]
Remove the whole `Path.Combine` call, and just use `Database.db` or `.\Database.db`.

**Reply by Bondarenko1990** [reply]
@huoyaoyuan 
![image](https://user-images.githubusercontent.com/32612104/172827172-bf157a4d-3057-4db2-b364-9de20ea9881d.png)
And I have error 
`Exception thrown: 'Microsoft.Data.Sqlite.SqliteException' in System.Private.CoreLib.dll
An exception of type 'Microsoft.Data.Sqlite.SqliteException' occurred in System.Private.CoreLib.dll but was not handled in user code
SQLite Error 14: 'unable to open database file'.`

![image](https://user-images.githubusercontent.com/32612104/172827466-b689c83f-5c20-467c-bf3a-612db8100ca4.png)
database not created

**Comment by ghost1372** [comment]
@Bondarenko1990  mybe this can help you

https://github.com/microsoft/WinUI-Gallery/blob/7238109ef49a3fea0884db05671ed11b2cec1f5d/WinUIGallery/Common/SuspensionManager.cs#L81

**Reply by Bondarenko1990** [reply]
I cannot use StorageFolder Or ApplicationData in Class Library
![image](https://user-images.githubusercontent.com/32612104/172226000-f58244fa-a65a-473a-a2d9-c8727e43f137.png)
